gradio/event_queue.py

```diff
@@ -0,0 +1,276 @@
+    DURATION_HISTORY = []
+    # When there is no estimation is calculated, default estimation is 1 sec
+    AVG_PROCESS_TIME = 1
+    AVG_CONCURRENT_PROCESS_TIME = 1
+    QUEUE_DURATION = 1
+    LIVE_QUEUE_UPDATES = True
+    SLEEP_WHEN_FREE = 0.001
+
+    @classmethod
+    def configure_queue(
+        cls,
+        server_path: str,
+        live_queue_updates=True,
+        queue_concurrency_count: int = 1,
+        data_gathering_start: int = 30,
+        update_intervals: int = 5,
+        duration_history_size=100,
+    ):
+        """
+        See Blocks.launch() docstring for the explanation of parameters.
+        """
+
+        if live_queue_updates is False and update_intervals == 5:
+            update_intervals = 10
+        cls.SERVER_PATH = server_path
+        cls.LIVE_QUEUE_UPDATES = live_queue_updates
+        cls.MAX_THREAD_COUNT = queue_concurrency_count
+        cls.DATA_GATHERING_STARTS_AT = data_gathering_start
+        cls.UPDATE_INTERVALS = update_intervals
+        cls.DURATION_HISTORY_SIZE = duration_history_size
+        cls.ACTIVE_JOBS = [None] * cls.MAX_THREAD_COUNT
+
+    @classmethod
+    async def init(
+        cls,
+    ) -> None:
+        run_coro_in_background(Queue.notify_clients)
+        run_coro_in_background(Queue.start_processing)
+
+    @classmethod
+    def close(cls):
+        cls.STOP = True
+
+    @classmethod
+    def resume(cls):
+        cls.STOP = False
+
+    # TODO: Remove prints
+    @classmethod
+    async def start_processing(cls) -> None:
+        while not cls.STOP:
+            if not cls.EVENT_QUEUE:
+                await asyncio.sleep(cls.SLEEP_WHEN_FREE)
+                continue
+
+            print(f"Searching for inactive job slots")
+            if not (None in cls.ACTIVE_JOBS):
+                await asyncio.sleep(1)
+                continue
+
+            # Using mutex to avoid editing a list in use
+            async with cls.LOCK:
+                print(f"Start Processing, found inactive job slot, now popping event")
+                event = cls.EVENT_QUEUE.pop(0)
+                print(f"Start Processing, found event, popped event: {event}")
+
+            cls.ACTIVE_JOBS[cls.ACTIVE_JOBS.index(None)] = event
+            run_coro_in_background(cls.process_event, event)
+            run_coro_in_background(cls.gather_data_for_first_ranks)
+            if cls.LIVE_QUEUE_UPDATES:
+                run_coro_in_background(cls.broadcast_estimation)
+
+    @classmethod
+    def push(cls, event: Event):
+        cls.EVENT_QUEUE.append(event)
+
+    @classmethod
+    def clean_job(cls, event: Event):
+        cls.ACTIVE_JOBS[cls.ACTIVE_JOBS.index(event)] = None
+
+    @classmethod
+    async def clean_event(cls, event: Event) -> None:
+        # Using mutex to avoid editing a list in use
+        async with cls.LOCK:
+            # TODO: Might log number of cleaned events in the future
+            try:  # TODO: Might search with event hash to speed up
```

Comment:
insstead of using try / catch, can you just just use an if statement? If event in queue...